api/src/main/java/edu/cornell/mannlib/vitro/webapp/controller/authenticate/ForgotPasswordController.java

```
@@ -0,0 +1,400 @@
+
+    private static final int DAYS_TO_USE_PASSWORD_LINK = 5;
+
+    private static final String TEMPLATE_NAME = "userAccounts-resetPasswordRequest.ftl";
+    private static final Log log = LogFactory.getLog(ForgotPasswordController.class.getName());
+    private CaptchaImplementation captchaImpl;
+
+    @Override
+    protected ResponseValues processRequest(VitroRequest vreq) throws Exception {
+        // Random time interval sleep so attacker can't calculate whether provided email is bound to an account or not
+        sleepForRandomTime();
+
+        captchaImpl = CaptchaServiceBean.getCaptchaImpl();
+
+        Map<String, Object> dataContext = new HashMap<>();
+        setCommonValues(dataContext, vreq);
+        CaptchaServiceBean.addCaptchaRelatedFieldsToPageContext(dataContext);
+        boolean isEnabled = isFunctionalityEnabled();
+        dataContext.put("isEnabled", isEnabled);
+
+        // Handle GET request (display the form) or print error message if functionality is disabled
+        if (!isEnabled || vreq.getMethod().equalsIgnoreCase("GET")) {
+            return showForm(dataContext);
+        }
+
+        return handlePostRequest(vreq, dataContext);
+    }
+
+    /**
+     * Handles a POST request for password reset. Validates the captcha, checks for spam mitigation,
+     * processes the password change request, and notifies the user.
+     *
+     * @param vreq        The Vitro request object.
+     * @param dataContext A map containing additional data context for processing the request.
+     * @return A response containing the appropriate values for rendering the result.
+     */
+    private ResponseValues handlePostRequest(VitroRequest vreq, Map<String, Object> dataContext) {
+        UserAccountsDao userAccountsDao = constructUserAccountsDao();
+        I18nBundle i18n = I18n.bundle(vreq);
+
+        String email = getNonNullTrimmedParameterValue(vreq, "email");
```

Please limit input parameters length to some reasonable sizes.

---

Should exceeding the limit when parsing arguments return an error or just in an empty string?

---

I would return an error.

---

SHould check be performed on backend and frontend or on the backend only?

---

I meant that it is better to check input variables lenght on the backend before processing.
On frontend maxlength input attribute can be used for input fields like email.

---

I did as you advised, but it is a redundant check as input type="email" checks for length as well :slightly_smiling_face:, nevertheless it shouldn't hurt.

---

Maybe first check if string is shorter than 320, and trim only in this case?